streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -87,9 +88,13 @@ int maybeCommit() {
      */
     int process() {
         int processed = 0;
-        final Iterator<Map.Entry<TaskId, StreamTask>> it = running.entrySet().iterator();
-        while (it.hasNext()) {
-            final StreamTask task = it.next().getValue();
+
+        // only iterate over tasks who has data in all its input topic partitions
+        final Map<TaskId, StreamTask> processable = running.entrySet().stream()
+                .filter(entry -> entry.getValue().isProcessable())
```

Comment:
The key is that we want to call `task.IsProcessable()` only once within each StreamThread iteration, since calling it N times will cause it to be forced to process. On the other hand, I'm not sure of the performance impact of this line either, so if there are any better ideas on achieving the following:

1. We only check `task.IsProcessable()` once within the iteration, after polled records have been put into the queue.

2. During processing, if a task is no longer "buffered on all partitions", we stop iterating it but still looping over others.

Reply:
I'm not sure I follow; putting `(!task.isProcessable()) continue;` in between lines 97 and 98 should be exactly the same as filtering this map before the loop.

Reply:
Ah, I'm confusing myself with part II and part III PRs :) Yes in this PR it is the same, but in https://github.com/apache/kafka/pull/5428 I did some further modifications so that `isProcessable` is only called once per iteration. The reason I did not put that into this PR is only because I am still investigating some unexpected perf diffs in that PR, and for this PR at least I can confirm that no perf diffs were introduced.

But since we have talked about it, I may just move that logic from 5428 to this one and see if it stays the same.